0018 - Add ADR about Horizontal Migration (#26)
PrestaShop/ADR: 0018-horizontal-migration.md

# 18. Horizontal migration

Date: 2022-03-24

## Status

Approved

## Context

In our current migration strategy, migrating a page to Symfony requires switching the view from Smarty to Twig, switching listings from HelperList to Grid, switching forms from HelperForm to Symfony forms, applying the CQRS layer, and more. 

> As a result of the migration, the Back office structure is heterogeneous, with some pages fully Symfony-based and others full legacy-based, increasing developer and performance overheaad. In addition, legacy subsistems must remain in place and continue being maintained, as long as there are legacy pages in PrestaShop. However, we still have year's worth of work ahead of us.

Reviewer: ```suggestion
As a result of the migration, the Back office structure is heterogeneous, with some pages fully Symfony-based and others full legacy-based, increasing developer and performance overhead. In addition, legacy subsystems must remain in place and continue being maintained, as long as there are legacy pages in PrestaShop. However, we still have year's worth of work ahead of us.
```

### First stage: migrate the "Controller layer"

The first stage consists in migrating the Controller Layer to Symfony (based on FrameworkBundleAdminControllers, services, sf router, and action-methods), but keeping most of the original behavior as it is, within the controller itself. This includes maintaining the Smarty template engine as well as closely related legacy components like HelperList, HelperForm, and others. 

A new namespace `Bridge` will be created within the PrestaShopBundle to contain intermediary code that will be removed as the migration progresses. For example, temporary traits will be added into horizontally migrated Syfony controllers, enabling any necessary legacy features which do not belong in a fully migrated Symfony controller (like Smarty support). These traits are meant to be removed when the features will no longer be required.

The `LegacyControllerBridgeInterface` replaces the controller that is present in Context, providing methods needed by modules but that don't belong in Symfony controllers, including:

- `setMedia`
- `addCss`
- `addJs`
- `addJqueryPlugin`
- `addJqueryUI`

> - ...

Reviewer: and public fields that were originally defined in the legacy controller '$controllerName' `toolbar_buttons`, ...

Reviewer 2: Please suggest an edit

Reviewer: ```suggestion

It will also provide access to public fields that were accessible from the `AdminController` class to keep as much backward compatibility as possible:

- `php_self`
- `controller_name`
- `table`
- `meta_title`
- ...
```

# 18. Horizontal migration

Date: 2022-03-24

## Status

Approved

## Context

In our current migration strategy, "migrating a page to Symfony" is not just migration: it's actually a migration + refactoring. This includes switching the view from Smarty to Twig, switching listings from HelperList to Grid, switching forms from HelperForm to Symfony forms, applying the CQRS layer, and more. 

This process takes a very long time, and as a result, the Back office structure has become heterogeneous, with some pages fully Symfony-based and others full legacy-based, increasing developer and performance overhead. 

> In addition, legacy subsystems must remain in place and continue being maintained, they cannot be removed as long as there are legacy pages in PrestaShop. Ad we still have years' worth of work ahead of us to finish. Let's imagine that migrating the rest of the Back office to Symfony using the current strategy will take 4 years. This means that 2 years years from now, we will still have about 25% of the pages running on the legacy architecture. And because of it, we will still have to maintain AdminController, Dispatcher, overrides, etc for the whole time, until the migration is finished – 4 years from now.

Reviewer: ```suggestion
In addition, legacy subsystems must remain in place and continue being maintained, they cannot be removed as long as there are legacy pages in PrestaShop. Ad we still have years' worth of work ahead of us to finish. Let's imagine that migrating the rest of the Back office to Symfony using the current strategy will take 4 years. This means that 2 years from now, we will still have about 25% of the pages running on the legacy architecture. And because of it, we will still have to maintain AdminController, Dispatcher, overrides, etc for the whole time, until the migration is finished – 4 years from now.
```
